File: go/a7p/validate.go
package a7p

import (
	profedit "a7p-go/a7p/profedit"
	"fmt"

	protovalidate "github.com/bufbuild/protovalidate-go"
	validator "github.com/go-playground/validator/v10"
)

const SwitchesMaxCount = 4

// Validate checks the content using protovalidate.
func ValidateProto(payload *profedit.Payload) error {
	validator, err := protovalidate.New()
	if err != nil {
		return fmt.Errorf("failed to create validator: %w", err)
	}

	if err := validator.Validate(payload); err != nil {
		return fmt.Errorf("validation failed: %w", err)
	}
	return nil
}

func ValidateSpec(payload *profedit.Payload) error {
	pld := convertPayload(payload)

	validate := validator.New()

	if err := validate.Struct(pld); err != nil {
		// t.Errorf("validation failed: %s", err)
		for _, e := range err.(validator.ValidationErrors) {
			fmt.Printf("Field: %s | Tag: %s | Value: %v\n", e.Field(), e.Tag(), e.Value())
		}
		return err
	}
	return nil
}

type swPosT struct {
	cIdx         int32
	reticleIdx   int32
	zoom         int32 `validate:"gte=1,lte=6"`
	distance     int32
	distanceFrom profedit.DType
}

type coefRowT struct {
	bcCd int32
	mv   int32
}

type profileT struct {
	profileName   string `validate:"max=50"`
	cartridgeName string `validate:"max=50"`
	bulletName    string `validate:"max=50"`
	shortNameTop  string `validate:"max=8"`
	shortNameBot  string `validate:"max=8"`
	userNote      string `validate:"max=250"`

	zeroX    int32 `validate:"gte=-600000,lte=600000"`
	zeroY    int32 `validate:"gte=-600000,lte=600000"`
	scHeight int32 `validate:"gte=-5000,lte=5000"`
	rTwist   int32 `validate:"gte=0,lte=10000"`

	cMuzzleVelocity     int32 `validate:"gte=100,lte=30000"`
	cZeroTemperature    int32 `validate:"gte=-100,lte=100"`
	cTCoeff             int32 `validate:"gte=2,lte=3000"`
	cZeroDistanceIdx    int32 `validate:"gte=0,lte=200"`
	cZeroAirTemperature int32 `validate:"gte=-100,lte=100"`
	cZeroAirPressure    int32 `validate:"required"`
	cZeroAirHumidity    int32 `validate:"gte=0,lte=100"`
	cZeroWPitch         int32 `validate:"gte=-90,lte=90"`
	cZeroPTemperature   int32 `validate:"gte=-100,lte=100"`

	bDiameter int32 `validate:"gte=1,lte=65535"`
	bWeight   int32 `validate:"gte=10,lte=65535"`
	bLength   int32 `validate:"gte=1,lte=10000"`

	twistDir   profedit.TwistDir `validate:""`
	bcType     profedit.GType    `validate:""`
	switches   []*swPosT         `validate:"required,min=1,max=4"`
	distances  []int32           `validate:"required,min=4,max=200"`
	coefRows   []*coefRowT       `validate:"required,min=1,max=200"`
	caliber    string            `validate:"max=50"`
	deviceUuid string            `validate:"max=50"`
}

type payloadT struct {
	profile *profileT `validate:"required"`
}

//

func convertPayload(payload *profedit.Payload) payloadT {
	return payloadT{profile: convertProfile(payload.Profile)}
}

func convertProfile(profile *profedit.Profile) *profileT {
	return &profileT{
		profileName:         profile.ProfileName,
		cartridgeName:       profile.CartridgeName,
		bulletName:          profile.BulletName,
		shortNameTop:        profile.ShortNameTop,
		shortNameBot:        profile.ShortNameBot,
		userNote:            profile.UserNote,
		zeroX:               profile.ZeroX,
		zeroY:               profile.ZeroY,
		scHeight:            profile.ScHeight,
		rTwist:              profile.RTwist,
		cMuzzleVelocity:     profile.CMuzzleVelocity,
		cZeroTemperature:    profile.CZeroTemperature,
		cTCoeff:             profile.CTCoeff,
		cZeroDistanceIdx:    profile.CZeroDistanceIdx,
		cZeroAirTemperature: profile.CZeroAirTemperature,
		cZeroAirPressure:    profile.CZeroAirPressure,
		cZeroAirHumidity:    profile.CZeroAirHumidity,
		cZeroWPitch:         profile.CZeroWPitch,
		cZeroPTemperature:   profile.CZeroPTemperature,
		bDiameter:           profile.BDiameter,
		bWeight:             profile.BWeight,
		bLength:             profile.BLength,
		twistDir:            profile.TwistDir,
		bcType:              profile.BcType,
		switches:            convertSwPos(profile.Switches),
		distances:           profile.Distances,
		coefRows:            convertCoefRows(profile.CoefRows),
		caliber:             profile.Caliber,
		deviceUuid:          profile.DeviceUuid,
	}
}

// Convert nested types
func convertSwPos(switches []*profedit.SwPos) []*swPosT {
	var result []*swPosT
	for _, s := range switches {
		result = append(result, &swPosT{
			cIdx:         s.CIdx,
			reticleIdx:   s.ReticleIdx,
			zoom:         s.Zoom,
			distance:     s.Distance,
			distanceFrom: s.DistanceFrom,
		})
	}
	return result
}

func convertCoefRows(coefRows []*profedit.CoefRow) []*coefRowT {
	var result []*coefRowT
	for _, r := range coefRows {
		result = append(result, &coefRowT{
			bcCd: r.BcCd,
			mv:   r.Mv,
		})
	}
	return result
}

File: js/src/validate.ts
import validateSchema from './generated/a7p_schema_validator.cjs';
import { CoefRow, Payload, Profile, SwitchProps } from './types.js';
import { ValidationError } from './errors.js';

// Converts the Payload (camelCase, matching the ts-proto/protobuf field
// names) into the plain snake_case JSON shape schema/a7p.schema.json is
// written against -- the same shape MessageToDict(preserving_proto_field_name=True)
// produces in py, and the manual conversion a7p_validator.dart does for dart.
function payloadToSchemaJson(payload: Payload): unknown {
    const p: Profile = payload.profile;
    return {
        profile: {
            profile_name: p.profileName,
            cartridge_name: p.cartridgeName,
            bullet_name: p.bulletName,
            short_name_top: p.shortNameTop,
            short_name_bot: p.shortNameBot,
            caliber: p.caliber,
            device_uuid: p.deviceUuid,
            user_note: p.userNote,
            zero_x: p.zeroX,
            zero_y: p.zeroY,
            sc_height: p.scHeight,
            r_twist: p.rTwist,
            twist_dir: p.twistDir,
            c_muzzle_velocity: p.cMuzzleVelocity,
            c_zero_temperature: p.cZeroTemperature,
            c_t_coeff: p.cTCoeff,
            c_zero_distance_idx: p.cZeroDistanceIdx,
            c_zero_air_temperature: p.cZeroAirTemperature,
            c_zero_air_pressure: p.cZeroAirPressure,
            c_zero_air_humidity: p.cZeroAirHumidity,
            c_zero_w_pitch: p.cZeroWPitch,
            c_zero_p_temperature: p.cZeroPTemperature,
            b_diameter: p.bDiameter,
            b_weight: p.bWeight,
            b_length: p.bLength,
            bc_type: p.bcType,
            switches: (p.switches ?? []).map((sw: SwitchProps) => ({
                c_idx: sw.cIdx,
                distance_from: sw.distanceFrom,
                distance: sw.distance,
                reticle_idx: sw.reticleIdx,
                zoom: sw.zoom,
            })),
            distances: p.distances,
            coef_rows: (p.coefRows ?? []).map((r: CoefRow) => ({
                bc_cd: r.bcCd,
                mv: r.mv,
            })),
        },
    };
}

// ajv reports errors as JSON Pointers rooted at the whole payload (e.g.
// "/profile/switches/0/c_idx"); strip the "/profile" prefix so messages
// read the same as the old yup-based errors did.
function fieldPath(instancePath: string): string {
    let path = instancePath;
    if (path.startsWith('/profile')) {
        path = path.slice('/profile'.length);
    }
    if (path.startsWith('/')) {
        path = path.slice(1);
    }
    return path.length === 0 ? 'payload' : path;
}

// Not expressible in plain JSON Schema (see coef_rows.x-unique-except-zero
// in schema/a7p.schema.json): all 'mv' values must be unique except that
// any number of rows may have mv == 0. Checked separately, same as py's
// schema_validator.py and dart's A7pValidator.
function uniqueMvError(rows: CoefRow[]): string | null {
    const seen = new Set<number>();
    for (const r of rows) {
        if (r.mv !== 0) {
            if (seen.has(r.mv)) {
                return "coef_rows: 'mv' values must be unique, except for mv == 0";
            }
            seen.add(r.mv);
        }
    }
    return null;
}

export const validate = (data: Payload, abortEarly: boolean = false): void => {
    if (!data.profile) {
        throw new ValidationError(['payload: missing profile']);
    }

    const messages: string[] = [];

    const ok = validateSchema(payloadToSchemaJson(data));
    if (!ok) {
        for (const e of validateSchema.errors ?? []) {
            messages.push(`${fieldPath(e.instancePath)}: ${e.message}`);
        }
    }

    const mvError = uniqueMvError(data.profile.coefRows ?? []);
    if (mvError) messages.push(mvError);

    if (messages.length > 0) {
        throw new ValidationError(abortEarly ? [messages[0]] : messages);
    }
};
